### DISCUSSION TITLE
Support for additional MCP primitives in Theia

### DISCUSSION THREAD

**Discussion by dsp-ant** [eclipse-theia/theia#14655] [Improvements]
Based on https://github.com/modelcontextprotocol/docs/pull/90#issuecomment-2554563118 discussion with @JonasHelming 

It's great to see Theia implement MCP.

The protocol features additional primitives, Prompts, Resources, Sampling and Roots that would be great to have implement in Theia. Prompts are often mapped to `@` or `/` commands in an assistant panel, for example in Zed `/my-prompt` would call the prompt and insert it. Resources are often attached as files and the application decides how to handle them.

A bit more interesting for an IDE, but sadly not widely supported by MCP servers, is the idea of Roots. This is a facility allowing MCP hosts/clients such as Theia inform servers about open workspaces, such that MCP servers can focus their work on this. This can be very useful for agentic MCP servers that must operate in a directory open in Theia. Sampling could be an interesting facility if MCP servers should be able to use LLMs without bringing their own mode SDK (e,g, openai or anthropic sdk).

These are just some ideas. Would be great to see the implementation becoming richer in capabiility.

**Comment by JonasHelming** [comment]
Thank you very much about the feedback!

Yes, we are interested in the other capabilities of MCP, too, let's break this down:

## Prompts
This is a no brainer. Theia AI (the underlying framework for the AI-powered Theia IDE) has already a flexible prompt management. In the Theia IDE, users can modify prompts, create new prompts and select the prompts for any agents. So the only thing we need to do really is to read in the prompts from MCP and make them part of our existing infrastructure. Do you have a good example for a MCP server providing prompts?

## Ressources
Seems also straight forward. If a server provides ressources, we could provide the generic resources access functions as tool functions and additionally we could provide a variable with a URI parameter to be used for explicitly referencing a sepcific resources (in the chat or a prompt)

## Roots
Yes, that sounds indeed very interesting and we could very easily expose roots to an MCP server, such as the current workspace, the current git origin or even something the user can decide on. What I did not see yet is: Are MCP server exposing whether they expect roots and what they do with it? Do you know a good example MCP server that uses roots?

## Sampling
Is there any server using this, yet? We could also implement this, we have integrations for openai, anthropic, ollama and llama-file, but it would be good to have a least one use case :-)

Thanks again for your input, we are really excited about the capabilities gained via MCP, great project for AI innovation!

**Reply by dsp-ant** [reply]
> This is a no brainer. Theia AI (the underlying framework for the AI-powered Theia IDE) has already a flexible prompt management. In the Theia IDE, users can modify prompts, create new prompts and select the prompts for any agents. So the only thing we need to do really is to read in the prompts from MCP and make them part of our existing infrastructure. Do you have a good example for a MCP server providing prompts?

https://github.com/modelcontextprotocol/servers/blob/main/src/sentry/src/mcp_server_sentry/server.py
and https://github.com/zed-industries/postgres-context-server/blob/main/index.mjs are examples. The Zed implementation also supports Autocompletion for prompts, since Zed's MCP server support it. It's a nice feature.

> Seems also straight forward. If a server provides ressources, we could provide the generic resources access functions as tool functions and additionally we could provide a variable with a URI parameter to be used for explicitly referencing a sepcific resources (in the chat or a prompt)

Using tools to access resources is quite nice. It is something we have been looking into as well. Other approaches to resources is 'attaching' them to the prompt. This is how Claude Desktop works. Since resources are 'application driven', it depends what makes most sense for Theia.

> Yes, that sounds indeed very interesting and we could very easily expose roots to an MCP server, such as the current workspace, the current git origin or even something the user can decide on. What I did not see yet is: Are MCP server exposing whether they expect roots and what they do with it? Do you know a good example MCP server that uses roots?

Clients advertise that they have roots. Servers use them if they want to. They are advisory. That being said, there is little use and if we discover we need stricter guarantees on the client side, we should consider making changes. There is no example server that uses roots since there is no client that effectively supports it, sadly. I build roots into the protocol because we had a bunch of people internally building MCP servers that require a notion where to operate on, but haven't gotten around actually implementing it in some of the IDEs we are using (including Zed)

> Is there any server using this, yet? We could also implement this, we have integrations for openai, anthropic, ollama and llama-file, but it would be good to have a least one use case :-)
> 
> Thanks again for your input, we are really excited about the capabilities gained via MCP, great project for AI innovation!

We have internal servers that do it. They are both clients and hosts and are effectively an agent loop using MCP servers as tools and providing sampling to them. There is also no client just yet. I totally see why most skip sampling and roots, but I think people are sleeping on what this can enable.

Thank you so much for all the work on Theia.

**Reply by JonasHelming** [reply]
> > This is a no brainer. Theia AI (the underlying framework for the AI-powered Theia IDE) has already a flexible prompt management. In the Theia IDE, users can modify prompts, create new prompts and select the prompts for any agents. So the only thing we need to do really is to read in the prompts from MCP and make them part of our existing infrastructure. Do you have a good example for a MCP server providing prompts?
> 
> https://github.com/modelcontextprotocol/servers/blob/main/src/sentry/src/mcp_server_sentry/server.py and https://github.com/zed-industries/postgres-context-server/blob/main/index.mjs are examples. The Zed implementation also supports Autocompletion for prompts, since Zed's MCP server support it. It's a nice feature.

Cool, I will look into this!

> 
> > Seems also straight forward. If a server provides ressources, we could provide the generic resources access functions as tool functions and additionally we could provide a variable with a URI parameter to be used for explicitly referencing a sepcific resources (in the chat or a prompt)
> 
> Using tools to access resources is quite nice. It is something we have been looking into as well. Other approaches to resources is 'attaching' them to the prompt. This is how Claude Desktop works. Since resources are 'application driven', it depends what makes most sense for Theia.

I think we have the same understanding here. Variables in Theia are the way we "attach something to the prompt" either in a system prompt or a user query. So I think we can just support both patterns easily!

> 
> > Yes, that sounds indeed very interesting and we could very easily expose roots to an MCP server, such as the current workspace, the current git origin or even something the user can decide on. What I did not see yet is: Are MCP server exposing whether they expect roots and what they do with it? Do you know a good example MCP server that uses roots?
> 
> Clients advertise that they have roots. Servers use them if they want to. They are advisory. That being said, there is little use and if we discover we need stricter guarantees on the client side, we should consider making changes. There is no example server that uses roots since there is no client that effectively supports it, sadly. I build roots into the protocol because we had a bunch of people internally building MCP servers that require a notion where to operate on, but haven't gotten around actually implementing it in some of the IDEs we are using (including Zed)

Needless to say that it would be super interesting to know what these servers do :-)
I agree with you, it's interesting, but a clear chicken and egg problem. We do not even have something to test this with if we do not create a MCP server for testing only
Maybe one suggestion: Wouldn't the [Filesystem MCP Server](https://github.com/modelcontextprotocol/servers/tree/main/src/filesystem) benefit from roots, so it understands the workspace the users wants to operate it? Currently it has allowed directories as an argument, but "roots" could augment or replace this. As an example, roots could be automatically allowed and you could use relative paths within roots. Just a quick idea to maybe boost this feature with an example.


> 
> > Is there any server using this, yet? We could also implement this, we have integrations for openai, anthropic, ollama and llama-file, but it would be good to have a least one use case :-)
> > Thanks again for your input, we are really excited about the capabilities gained via MCP, great project for AI innovation!
> 
> We have internal servers that do it. They are both clients and hosts and are effectively an agent loop using MCP servers as tools and providing sampling to them. There is also no client just yet. I totally see why most skip sampling and roots, but I think people are sleeping on what this can enable.

Again a chicken an egg problem. This is also overlapping with the concept of agents that Theia nativley has.
However of course it would be interesting to be extendable with MCP Servers, specifially if there are interesting implementations in the wild. The question to me is whether "sampling" is a good term long run, because based on what you describe it sounds actually like more than that. In my understanding it is providing the MCP Server with access to an LLM, so more like "Model Request" or "LLM Interactions". (just my two cents)

> Thank you so much for all the work on Theia.

Thank YOU for MCP :-) and thank you for sharing your thoughts here!

**Reply by dsp-ant** [reply]
> Maybe one suggestion: Wouldn't the [Filesystem MCP Server](https://github.com/modelcontextprotocol/servers/tree/main/src/filesystem) benefit from roots, so it understands the workspace the users wants to operate it? Currently it has allowed directories as an argument, but "roots" could augment or replace this. As an example, roots could be automatically allowed and you could use relative paths within roots. Just a quick idea to maybe boost this feature with an example.

Great suggestion.

**Reply by JonasHelming** [reply]
I reported this: https://github.com/modelcontextprotocol/servers/issues/401

**Reply by JonasHelming** [reply]
@dsp-ant Re "roots": Now we even have a feature request that requires "root": https://github.com/eclipse-theia/theia/issues/14689
:-)
Do you see a chance that https://github.com/modelcontextprotocol/servers/issues/401 could get fixed? (looks pretty small)
This would probably also make "roots" more commonly implemented and used.

**Reply by dsp-ant** [reply]
@JonasHelming I think we can do that. Feel free to remind me over bluesky, twitter or my email address if i get behind on this one. Ill see that i can make space for it.
